src/shrinkray/subprocess/protocol.py

```diff
@@ -59,6 +59,11 @@ class ProgressUpdate:
     test_output_preview: str = ""
     # Currently running test ID (None if no test running)
     active_test_id: int | None = None
+    # Return code of most recently completed test (None if no completed test yet)
+    last_test_return_code: int | None = None
```

Comment:
The fact that you are tracking return codes separately from output is _very_ suspicious, and is a bug waiting to happen if it's not wrong already.